src/FSharp.AWS.DynamoDB/TableContext.fs

```diff
@@ -980,24 +980,33 @@ type TableContext<'TRecord> internal
 type TableContext =
 
     /// <summary>
-    ///     Creates a DynamoDB client instance for given F# record and table name.
+    ///     Creates a DynamoDB client instance for given F# record and table name. <br/>
+    ///     See <c>CreateAsync</c> for the ability to create and/or verify the Table.
+    /// </summary>
+    /// <param name="client">DynamoDB client instance.</param>
+    /// <param name="tableName">Table name to target.</param>
+    /// <param name="metricsCollector">Function to receive request metrics.</param>
+    static member CreateUnverified<'TRecord>(client : IAmazonDynamoDB, tableName : string, ?metricsCollector : RequestMetrics -> unit) : TableContext<'TRecord> =
```

Comment:
@samritchie Not sure how you feel about this, but in general I would try to offer a synchronous/lazy creation method like this - while the default behavior of doing the verification is desirable as a reasonable default, I'd expect table creation to be managed by the deploy process in general.

This is also important for serverless scenarios, where you want to be confident no avoidable processing is potentially taking place on a per-request basis. 

Assuming you're OK with introducing it, I can also shift this into a separate PR rather than jamming it in here.

Reply:
Personally I don’t use verification (or automatic table creation) - I've found it is too hard to manage blue/green deployments with online index/table changes & believe this is better handled by application-level feature toggles. That doesn’t mean it’s right for everyone, but I don’t have a problem in principle with the change.

The developer ergonomics are a little awkward though. Perhaps it would be better to split creation/verification into distinct methods on TableContext, something like:

```
do! table.CreateIfNotExistsAsync(provisionedThroughput)
// or 
do! table.VerifyAsync()
```

And at that point I have to wonder whether we’re better off just using a public constructor for the TableContext - `Create` is a fairly ambiguous name for the factory method.

I do think a lot of (possibly most) people will be using the sync-over-async factory and taking the blocking hit on app startup so we'll likely need to leave the `Create` methods in for now.

I'm open to any other suggestions there...

Reply:
Another important semantic to be able to offer is to be able to 'put' a provisioned throughput level, either as part of creation (first time), or by visiting it and updating. I've implemented it like so: https://github.com/jet/equinox/blob/ddb/src/Equinox.DynamoStore/DynamoStore.fs#L471-L482

Cosmos doesn't provide such a facility directly either: https://github.com/jet/equinox/blob/ddb/src/Equinox.CosmosStore/CosmosStore.fs#L543-L555

In the context of Equinox, I have logic for one time checks and/or initialization, which is currently used to ensure a supporting stored procedure is in place.

----

For better or for worse, the current `Create` method is ideal for interactive scenarios and/or hacking about (e.g. I was able to [flip two flags and get stuff done](https://github.com/jet/equinox/blob/acccec32fabb27ec03b731a9a1c340fb341e3ae7/src/Equinox.DynamoStore/DynamoStore.fs#L340) in advance of wiring up more thorough initialization logic. Perhaps it should could/should move into the `namespace Scripting` (do you think an obsoleted Create method there with a message about doing `open Scripting` or migrating to async semantic methods makes sense or do we trust people to work it out/ suffer in pain without getting annoyed) ?

My thinking in naming it `CreateUnverified` was to make people think about whether that really is the best thing for then them to do in their system - having the default behavior be to verify the schema is pretty important IMO.

My original thinking/design was to be able to type `TableContext.Create` and have the completion list look approx like this, (but I only went half way in the name of backcompat):
- Create (the existing one)
- CreateIfNotExists (but that's called CreateAsync atm and has some messy semantics and defaulting)
- CreateUnverified

---

Setting aside all.backcompat concerns I guess what you've suggested makes sense. I propose the names as follows:

- remove `Create` (see Initialize for replacement - arguably it should be called `InitializeSynchronously`)
- remove `CreateAsync` (has ambiguous throughput param and strange defaulting)
- metricsCollector now only on present on and/or documented via ctor and Initialize
- shift `Create` to Scripting, leave the 10,10 provisioning default, forward all ctor args to ctor

```
type TableContext<TRecord>()    

    /// NOTE: Does not verify or initialize the table (see InitializeTableAsync and VerifyTableAsync APIs)
    new <'TRecord>(client : IAmazonDynamoDB, tableName : string, ?metricsCollector : RequestMetrics -> unit) : TableContext<'TRecord>

    /// Verifies the table is present and has the correct schema, or throws 
    /// See also InitializeTableAsync, which can be 
    member _.VerifyTableAsync() : Async<unit>

    /// Ensures the table has been created and indexes have been applied.
    /// updateThroughputIfExists: applies the specified throughput value, even if it already exists. Default: False -- leave it as currently configured if the table is already present

    // NOTE: no defaulting of throughput anymore
    member _.InitializeTableAsync(throughput : Throughput, ?updateThroughputIfExists : bool = false) : Async<unit>
```

As new provisioning modes are added, `throughput` could become a DU. 

That leaves the question of how to provide something that `Create` calls can migrate to.

```
module Scripting =

  type InitializationMode =
      | VerifyOnly
      | Create of provisionedThroughput : ProvisionedThroughput
      | CreateOrReprovision  of provisionedThroughput : ProvisionedThroughput
  
  type TableContext with

        // No defaulting of throughput, 
        static member Initialize<'TRecord>
            (   client : IAmazonDynamoDB, tableName : string,
                 ?mode : InitMode, // default: VerifyOnly
                 ?metricsCollector : RequestMetrics -> unit) : TableContext<'TRecord>
                                  
```

And I'd leave the Stub Table.Create for the moment:

```
    [<Obsolete("Please use either 1. TableContext constructor (followed by VerifyTableAsync or InitializeTableAsync) OR 2. Scripting.TableContext.Initialize")>]
    /// impl 
    static member Create<'TRecord>
        (   client : IAmazonDynamoDB, tableName : string,
             ?verifyTable = true, ?createIfNotExists = false,
             ?provisionedThroughput : ProvisionedThroughput, // defaults to 10,10
             ?metricsCollector : RequestMetrics -> unit) : TableContext<'TRecord>
```

Let me know if anything feels wrong - will make a PR shortly./

Reply:
> Another important semantic to be able to offer is to be able to 'put' a provisioned throughput level, either as part of creation (first time), or by visiting it and updating.

Maybe. The provisioned throughput support is very basic - it doesn’t cater for different throughput levels for different indices, or the on-demand/autoscaling modes as you point out. I also think it’s likely someone will unintentionally downgrade the throughput on a table or index. My view on this functionality is that it was more of a value-add to table creation (or more likely, Eirik didn’t want to hard-code a default without the ability to customise it) rather than a realistic attempt at implementing configuration as code.

I'd probably be more comfortable leaving provisioning updates on existing tables to explicit UpdateProvisionedThroughput calls - maybe we can cache the throughput used if the table was created to ensure it’s a no-op if run immediately after Initialize.

Reply:
It seems we are agreed on 
a) removing defaulting
b) confining specification of throughput to as few paths as possible - i.e. for autoscale and other more broadly managed provisioning rules, we don't want people to be requiring specific versions of `FSharp.AWS.DyanmoDB` etc and/or enable being able to route around it by doing the provisioning out of band, and only using `VerifyTableAsync`

My intent was to only have the reprovision facility be an opt-in flag on the `InitializeTableAsync` call - see  https://github.com/fsprojects/FSharp.AWS.DynamoDB/pull/43

It turns out that you get an exn with `The requested throughput value equals the current value` if you do an update not changing the values

If the feature was to be dropped:
a) I'd probably rename to `CreateTableIfNotExistsAsync`
b) I'd have to shift that logic into the Equinox provisioning tool (i.e. run a Create, followed by doing that logic)

Will ping when I have final code in there, and will let you make a call on it

One other escape hatch that may be worth providing might be to provide a case in the InitializationMode DU that lets you supply a `Create`/`UpdateTableRequest -> unit` function to do the manipulation in a custom manner.